{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## V2 Idea Time\n",
    "Read a passed .fna file for all its contents. This is a generalization of V1 that only read one passed block of <br> \n",
    "\\> fa header <br> \n",
    "then sequence AAGCGGGGT... <br>\n",
    "even if there were more. This new reader will still take a ```fragmentSize```, but it should read the file line by line until it reaches a new header or end of file. It should read in chunks of `fragmentSize`, automatically converting these chunks into 256 length vectors. I should probably store this in a numpy space because I will be working in 3D at this point. \n",
    "\n",
    "D1: \\> fa header <br>\n",
    "D2: tetramer vector header <br>\n",
    "D3: the number of `fragmentSize` chunks that fit into \\> fa header\n",
    "\n",
    "### Pseudocode\n",
    "1. Open file\n",
    "2. Readlines\n",
    "3. Check until first character of line is \"\\>\"\n",
    "4. Store the remainder of the line as variable/put as key of a dictionary\n",
    "    * Enter the code of V1 genome_to_tetramer pipeline (break when \\> is reached)\n",
    "    * Transform into a 2D matrix of fragments / 256 length vector\n",
    "    * Set dictionary key to this value\n",
    "5. Repeat until end of file\n",
    "6. Save resulting dictionary as a \"pickle\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "### Setup\n",
    "lenFragment = 1000\n",
    "genomeFragmentDict = {}\n",
    "fileName = 'vibrio_staph_ecoli_vibrio-aerogenes.fna'#\"GCF_000013425.1_ASM1342v1_genomic.fna(staph aureus)\"#\"Hodgkinia_cicadicola_tetund1_genome.fasta\"#\"Staph_argenteus_58113_genome.fasta\"\"Deinococcus_radiodurans_R1dM1_genome.fasta\"#\n",
    "fileOutput = 'multiTetraFreq.pickle'\n",
    "fastaHeader = ''\n",
    "\n",
    "tetramerDicRepository = {key:[] for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "tetramerDic = {key:0 for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fileName) as file:\n",
    "    for line in file.readlines():\n",
    "        if line[0] == '>':\n",
    "            fastaHeader = line[1:].strip()\n",
    "            genomeFragmentDict[fastaHeader] = []\n",
    "            fragment = \"\"\n",
    "            pass\n",
    "        else:\n",
    "            line = line.rstrip().upper()\n",
    "            # Lets keep this for now, it just keeps the edge cases where the last few characters\n",
    "            # aren't a multiple of 1000 from getting in, not an issue at the moment, will be an issue\n",
    "            # if I want to read fastqs (short reads)\n",
    "            if len(fragment) < lenFragment:\n",
    "                pass\n",
    "            else:\n",
    "                # this is where I should eventuall \n",
    "                # convert my string fragment into a tetramer dictionary\n",
    "                # but I just want something working right now\n",
    "                genomeFragmentDict[fastaHeader].append(fragment[:lenFragment])\n",
    "                fragment = fragment[lenFragment:]\n",
    "\n",
    "            fragment += line "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    },
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Till 0 run took 0.0002658367156982422\n",
      "Till 50 run took 0.01678013801574707\n",
      "Till 100 run took 0.014596700668334961\n",
      "Till 150 run took 0.014162778854370117\n",
      "Till 200 run took 0.015620231628417969\n",
      "Till 250 run took 0.015303850173950195\n",
      "Till 300 run took 0.015611410140991211\n",
      "Till 350 run took 0.01651453971862793\n",
      "Till 400 run took 0.01601696014404297\n",
      "Till 450 run took 0.014963388442993164\n",
      "Till 500 run took 0.015014171600341797\n",
      "Till 550 run took 0.014865636825561523\n",
      "Till 600 run took 0.014515161514282227\n",
      "Till 650 run took 0.014976263046264648\n",
      "Till 700 run took 0.01432037353515625\n",
      "Till 750 run took 0.01496124267578125\n",
      "Till 800 run took 0.015503644943237305\n",
      "Till 850 run took 0.015710830688476562\n",
      "Till 900 run took 0.015878677368164062\n",
      "Till 950 run took 0.01525259017944336\n",
      "Till 1000 run took 0.015709638595581055\n",
      "Till 1050 run took 0.015130996704101562\n",
      "Till 1100 run took 0.01489400863647461\n",
      "Till 1150 run took 0.01446223258972168\n",
      "Till 1200 run took 0.014118432998657227\n",
      "Till 1250 run took 0.012813568115234375\n",
      "Till 1300 run took 0.012494087219238281\n",
      "Till 1350 run took 0.012548208236694336\n",
      "Till 1400 run took 0.013154268264770508\n",
      "Till 1450 run took 0.01540994644165039\n",
      "Till 1500 run took 0.014868974685668945\n",
      "Till 1550 run took 0.014043569564819336\n",
      "Till 1600 run took 0.015671730041503906\n",
      "Till 1650 run took 0.013921499252319336\n",
      "Till 1700 run took 0.014972209930419922\n",
      "Till 1750 run took 0.015256881713867188\n",
      "Till 1800 run took 0.015610456466674805\n",
      "Till 1850 run took 0.01549220085144043\n",
      "Till 1900 run took 0.015293598175048828\n",
      "Till 1950 run took 0.015126943588256836\n",
      "Till 2000 run took 0.015340805053710938\n",
      "Till 2050 run took 0.016386985778808594\n",
      "Till 2100 run took 0.0159451961517334\n",
      "Till 2150 run took 0.01677560806274414\n",
      "Till 2200 run took 0.015466928482055664\n",
      "Till 2250 run took 0.013873100280761719\n",
      "Till 2300 run took 0.015418291091918945\n",
      "Till 2350 run took 0.015194177627563477\n",
      "Till 2400 run took 0.015313148498535156\n",
      "Till 2450 run took 0.015140771865844727\n",
      "Till 2500 run took 0.015482425689697266\n",
      "Till 2550 run took 0.015400409698486328\n",
      "Till 2600 run took 0.015677690505981445\n",
      "Till 2650 run took 0.015050888061523438\n",
      "Till 2700 run took 0.014860153198242188\n",
      "Till 2750 run took 0.01434469223022461\n",
      "Till 2800 run took 0.015465736389160156\n",
      "Till 2850 run took 0.01517629623413086\n",
      "Till 2900 run took 0.01609516143798828\n",
      "Till 2950 run took 0.013387441635131836\n",
      "Till 3000 run took 0.014226675033569336\n",
      "Till 3050 run took 0.015840768814086914\n",
      "Till 3100 run took 0.015206336975097656\n",
      "Till 3150 run took 0.014765262603759766\n",
      "Till 3200 run took 0.014687538146972656\n",
      "Till 3250 run took 0.0149993896484375\n",
      "Till 3300 run took 0.013992071151733398\n",
      "Till 3350 run took 0.013656854629516602\n",
      "Till 3400 run took 0.012476682662963867\n",
      "Till 3450 run took 0.012422800064086914\n",
      "Till 3500 run took 0.012353181838989258\n",
      "Till 3550 run took 0.013044595718383789\n",
      "Till 3600 run took 0.014501333236694336\n",
      "Till 3650 run took 0.015040874481201172\n",
      "Till 3700 run took 0.014596939086914062\n",
      "Till 3750 run took 0.012959480285644531\n",
      "Till 3800 run took 0.014712095260620117\n",
      "Till 3850 run took 0.014851570129394531\n",
      "Till 3900 run took 0.01499795913696289\n",
      "Till 3950 run took 0.015578269958496094\n",
      "Till 4000 run took 0.01558685302734375\n",
      "Till 4050 run took 0.015472173690795898\n",
      "Till 4100 run took 0.0155029296875\n",
      "Till 4150 run took 0.016626596450805664\n",
      "Till 4200 run took 0.014338254928588867\n",
      "Till 4250 run took 0.014683246612548828\n",
      "Till 4300 run took 0.015416860580444336\n",
      "Till 4350 run took 0.017033815383911133\n",
      "Till 4400 run took 0.014560937881469727\n",
      "Till 4450 run took 0.014632701873779297\n",
      "Till 4500 run took 0.014863252639770508\n",
      "Till 4550 run took 0.015888214111328125\n",
      "Till 4600 run took 0.015110492706298828\n",
      "done with NC_000913.3 Escherichia coli str. K-12 substr. MG1655, complete genome\n",
      "Till 0 run took 0.0003085136413574219\n",
      "Till 50 run took 0.015465736389160156\n",
      "Till 100 run took 0.016721248626708984\n",
      "Till 150 run took 0.016798973083496094\n",
      "Till 200 run took 0.017468929290771484\n",
      "Till 250 run took 0.01645374298095703\n",
      "Till 300 run took 0.016000747680664062\n",
      "Till 350 run took 0.015118598937988281\n",
      "Till 400 run took 0.017042875289916992\n",
      "Till 450 run took 0.01612234115600586\n",
      "Till 500 run took 0.015807628631591797\n",
      "Till 550 run took 0.015685558319091797\n",
      "Till 600 run took 0.014511346817016602\n",
      "Till 650 run took 0.013239383697509766\n",
      "Till 700 run took 0.01458597183227539\n",
      "Till 750 run took 0.015764236450195312\n",
      "Till 800 run took 0.016170501708984375\n",
      "Till 850 run took 0.015241622924804688\n",
      "Till 900 run took 0.015480995178222656\n",
      "Till 950 run took 0.014883756637573242\n",
      "Till 1000 run took 0.015049219131469727\n",
      "Till 1050 run took 0.014531612396240234\n",
      "Till 1100 run took 0.0128631591796875\n",
      "Till 1150 run took 0.012445688247680664\n",
      "Till 1200 run took 0.012238025665283203\n",
      "Till 1250 run took 0.01262354850769043\n",
      "Till 1300 run took 0.012548446655273438\n",
      "Till 1350 run took 0.012508869171142578\n",
      "Till 1400 run took 0.012435674667358398\n",
      "Till 1450 run took 0.013620853424072266\n",
      "Till 1500 run took 0.01412200927734375\n",
      "Till 1550 run took 0.013471364974975586\n",
      "Till 1600 run took 0.014300346374511719\n",
      "Till 1650 run took 0.0131988525390625\n",
      "Till 1700 run took 0.013022899627685547\n",
      "Till 1750 run took 0.014084577560424805\n",
      "Till 1800 run took 0.015007734298706055\n",
      "Till 1850 run took 0.014039754867553711\n",
      "Till 1900 run took 0.014574289321899414\n",
      "Till 1950 run took 0.015168190002441406\n",
      "Till 2000 run took 0.013538360595703125\n",
      "Till 2050 run took 0.013207674026489258\n",
      "Till 2100 run took 0.013093709945678711\n",
      "Till 2150 run took 0.012966632843017578\n",
      "Till 2200 run took 0.014441728591918945\n",
      "Till 2250 run took 0.013733625411987305\n",
      "Till 2300 run took 0.014804601669311523\n",
      "Till 2350 run took 0.014616966247558594\n",
      "Till 2400 run took 0.015660524368286133\n",
      "Till 2450 run took 0.015554428100585938\n",
      "Till 2500 run took 0.015702486038208008\n",
      "Till 2550 run took 0.015205860137939453\n",
      "Till 2600 run took 0.015094757080078125\n",
      "Till 2650 run took 0.014754295349121094\n",
      "Till 2700 run took 0.014369964599609375\n",
      "Till 2750 run took 0.01371145248413086\n",
      "Till 2800 run took 0.012725353240966797\n",
      "done with NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome\n",
      "Till 0 run took 0.0002734661102294922\n",
      "Till 50 run took 0.014019489288330078\n",
      "Till 100 run took 0.015220403671264648\n",
      "Till 150 run took 0.01571512222290039\n",
      "Till 200 run took 0.015184879302978516\n",
      "Till 250 run took 0.015473604202270508\n",
      "Till 300 run took 0.015539407730102539\n",
      "Till 350 run took 0.015240192413330078\n",
      "Till 400 run took 0.015550851821899414\n",
      "Till 450 run took 0.015472888946533203\n",
      "Till 500 run took 0.014877557754516602\n",
      "Till 550 run took 0.015755891799926758\n",
      "Till 600 run took 0.015324592590332031\n",
      "Till 650 run took 0.015765905380249023\n",
      "Till 700 run took 0.021370649337768555\n",
      "Till 750 run took 0.015307903289794922\n",
      "Till 800 run took 0.014200210571289062\n",
      "Till 850 run took 0.013110160827636719\n",
      "Till 900 run took 0.014867544174194336\n",
      "Till 950 run took 0.015901565551757812\n",
      "Till 1000 run took 0.016794204711914062\n",
      "Till 1050 run took 0.015465259552001953\n",
      "Till 1100 run took 0.015485286712646484\n",
      "Till 1150 run took 0.015069961547851562\n",
      "Till 1200 run took 0.014242172241210938\n",
      "Till 1250 run took 0.012917518615722656\n",
      "Till 1300 run took 0.012570381164550781\n",
      "Till 1350 run took 0.012599706649780273\n",
      "Till 1400 run took 0.012507200241088867\n",
      "Till 1450 run took 0.013309240341186523\n",
      "Till 1500 run took 0.013527870178222656\n",
      "Till 1550 run took 0.013362884521484375\n",
      "Till 1600 run took 0.015147924423217773\n",
      "Till 1650 run took 0.013997554779052734\n",
      "Till 1700 run took 0.01585698127746582\n",
      "Till 1750 run took 0.015100717544555664\n",
      "Till 1800 run took 0.015766143798828125\n",
      "Till 1850 run took 0.015363216400146484\n",
      "Till 1900 run took 0.014324665069580078\n",
      "Till 1950 run took 0.01503443717956543\n",
      "Till 2000 run took 0.015188932418823242\n",
      "Till 2050 run took 0.015357732772827148\n",
      "Till 2100 run took 0.015397787094116211\n",
      "Till 2150 run took 0.015274763107299805\n",
      "Till 2200 run took 0.015356302261352539\n",
      "Till 2250 run took 0.014446020126342773\n",
      "Till 2300 run took 0.013121366500854492\n",
      "Till 2350 run took 0.012940645217895508\n",
      "Till 2400 run took 0.014940977096557617\n",
      "Till 2450 run took 0.013113021850585938\n",
      "Till 2500 run took 0.012501716613769531\n",
      "Till 2550 run took 0.01242208480834961\n",
      "Till 2600 run took 0.013036727905273438\n",
      "Till 2650 run took 0.01258397102355957\n",
      "Till 2700 run took 0.012530326843261719\n",
      "Till 2750 run took 0.012555837631225586\n",
      "Till 2800 run took 0.012611865997314453\n",
      "Till 2850 run took 0.01275014877319336\n",
      "Till 2900 run took 0.013345003128051758\n",
      "done with NZ_CP043554.1 Vibrio cholerae strain RFB16 chromosome 1, complete sequence\n",
      "Till 0 run took 0.0002827644348144531\n",
      "Till 50 run took 0.014173030853271484\n",
      "Till 100 run took 0.015065670013427734\n",
      "Till 150 run took 0.015103578567504883\n",
      "Till 200 run took 0.015351533889770508\n",
      "Till 250 run took 0.014777183532714844\n",
      "Till 300 run took 0.014527559280395508\n",
      "Till 350 run took 0.013430595397949219\n",
      "Till 400 run took 0.012395858764648438\n",
      "Till 450 run took 0.012359142303466797\n",
      "Till 500 run took 0.01310276985168457\n",
      "Till 550 run took 0.01470041275024414\n",
      "Till 600 run took 0.014327526092529297\n",
      "Till 650 run took 0.013883352279663086\n",
      "Till 700 run took 0.012514591217041016\n",
      "Till 750 run took 0.015077829360961914\n",
      "Till 800 run took 0.01579451560974121\n",
      "Till 850 run took 0.015444755554199219\n",
      "Till 900 run took 0.015841960906982422\n",
      "Till 950 run took 0.015493392944335938\n",
      "Till 1000 run took 0.015814542770385742\n",
      "Till 1050 run took 0.014732122421264648\n",
      "Till 1100 run took 0.015349388122558594\n",
      "done with NZ_CP043556.1 Vibrio cholerae strain RFB16 chromosome 2, complete sequence\n",
      "Till 0 run took 0.0002696514129638672\n",
      "done with NZ_CP043555.1 Vibrio cholerae strain RFB16 plasmid unnamed, complete sequence\n",
      "Till 0 run took 0.00038170814514160156\n",
      "done with NZ_FQXZ01000077.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "done with NZ_FQXZ01000076.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "done with NZ_FQXZ01000073.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003466606140136719\n",
      "done with NZ_FQXZ01000071.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003314018249511719\n",
      "done with NZ_FQXZ01000070.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003540515899658203\n",
      "done with NZ_FQXZ01000068.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00034427642822265625\n",
      "done with NZ_FQXZ01000067.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003190040588378906\n",
      "done with NZ_FQXZ01000066.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003783702850341797\n",
      "done with NZ_FQXZ01000062.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003056526184082031\n",
      "done with NZ_FQXZ01000061.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003809928894042969\n",
      "done with NZ_FQXZ01000060.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0004031658172607422\n",
      "done with NZ_FQXZ01000056.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003578662872314453\n",
      "done with NZ_FQXZ01000072.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00036454200744628906\n",
      "done with NZ_FQXZ01000053.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003275871276855469\n",
      "done with NZ_FQXZ01000052.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003142356872558594\n",
      "done with NZ_FQXZ01000048.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003018379211425781\n",
      "done with NZ_FQXZ01000047.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00028133392333984375\n",
      "done with NZ_FQXZ01000063.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00031685829162597656\n",
      "done with NZ_FQXZ01000020.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00035262107849121094\n",
      "Till 50 run took 0.015459537506103516\n",
      "done with NZ_FQXZ01000017.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00028133392333984375\n",
      "done with NZ_FQXZ01000028.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002734661102294922\n",
      "done with NZ_FQXZ01000031.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003070831298828125\n",
      "done with NZ_FQXZ01000069.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003113746643066406\n",
      "done with NZ_FQXZ01000050.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002810955047607422\n",
      "Till 50 run took 0.015378713607788086\n",
      "Till 100 run took 0.016215801239013672\n",
      "Till 150 run took 0.016356945037841797\n",
      "done with NZ_FQXZ01000014.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003376007080078125\n",
      "done with NZ_FQXZ01000054.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003514289855957031\n",
      "Till 50 run took 0.015858173370361328\n",
      "done with NZ_FQXZ01000001.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00031876564025878906\n",
      "Till 50 run took 0.014802694320678711\n",
      "Till 100 run took 0.015168428421020508\n",
      "Till 150 run took 0.015111923217773438\n",
      "Till 200 run took 0.015150070190429688\n",
      "Till 250 run took 0.01473379135131836\n",
      "done with NZ_FQXZ01000005.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003135204315185547\n",
      "Till 50 run took 0.014586448669433594\n",
      "Till 100 run took 0.013029813766479492\n",
      "done with NZ_FQXZ01000035.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "done with NZ_FQXZ01000074.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00029277801513671875\n",
      "done with NZ_FQXZ01000059.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002925395965576172\n",
      "done with NZ_FQXZ01000002.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00029158592224121094\n",
      "Till 50 run took 0.015714645385742188\n",
      "Till 100 run took 0.015038490295410156\n",
      "done with NZ_FQXZ01000023.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002627372741699219\n",
      "done with NZ_FQXZ01000064.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002675056457519531\n",
      "Till 50 run took 0.014691352844238281\n",
      "Till 100 run took 0.015068292617797852\n",
      "done with NZ_FQXZ01000032.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00032210350036621094\n",
      "Till 50 run took 0.015116214752197266\n",
      "done with NZ_FQXZ01000041.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003046989440917969\n",
      "done with NZ_FQXZ01000049.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002739429473876953\n",
      "Till 50 run took 0.015569448471069336\n",
      "done with NZ_FQXZ01000004.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "done with NZ_FQXZ01000058.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003387928009033203\n",
      "Till 50 run took 0.015579700469970703\n",
      "Till 100 run took 0.015349388122558594\n",
      "done with NZ_FQXZ01000013.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003039836883544922\n",
      "Till 50 run took 0.014818429946899414\n",
      "Till 100 run took 0.014955759048461914\n",
      "Till 150 run took 0.015215396881103516\n",
      "Till 200 run took 0.016182422637939453\n",
      "Till 250 run took 0.015511274337768555\n",
      "Till 300 run took 0.015481710433959961\n",
      "Till 350 run took 0.015839338302612305\n",
      "done with NZ_FQXZ01000007.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002741813659667969\n",
      "Till 50 run took 0.014830589294433594\n",
      "done with NZ_FQXZ01000027.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003132820129394531\n",
      "Till 50 run took 0.015266895294189453\n",
      "done with NZ_FQXZ01000026.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002541542053222656\n",
      "Till 50 run took 0.014032125473022461\n",
      "done with NZ_FQXZ01000003.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002639293670654297\n",
      "Till 50 run took 0.015810012817382812\n",
      "Till 100 run took 0.016263246536254883\n",
      "done with NZ_FQXZ01000006.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00031447410583496094\n",
      "Till 50 run took 0.014886140823364258\n",
      "done with NZ_FQXZ01000011.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003018379211425781\n",
      "done with NZ_FQXZ01000016.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00026154518127441406\n",
      "Till 50 run took 0.012598037719726562\n",
      "Till 100 run took 0.014935016632080078\n",
      "Till 150 run took 0.015050888061523438\n",
      "Till 200 run took 0.015480995178222656\n",
      "Till 250 run took 0.015382528305053711\n",
      "Till 300 run took 0.01506495475769043\n",
      "done with NZ_FQXZ01000039.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002543926239013672\n",
      "Till 50 run took 0.01246333122253418\n",
      "done with NZ_FQXZ01000018.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002582073211669922\n",
      "done with NZ_FQXZ01000010.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002799034118652344\n",
      "Till 50 run took 0.013956785202026367\n",
      "Till 100 run took 0.015179157257080078\n",
      "Till 150 run took 0.015236854553222656\n",
      "Till 200 run took 0.014986038208007812\n",
      "Till 250 run took 0.015378713607788086\n",
      "Till 300 run took 0.015072107315063477\n",
      "Till 350 run took 0.014910697937011719\n",
      "Till 400 run took 0.01514577865600586\n",
      "done with NZ_FQXZ01000046.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002765655517578125\n",
      "Till 50 run took 0.013274192810058594\n",
      "done with NZ_FQXZ01000022.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00026416778564453125\n",
      "Till 50 run took 0.012799263000488281\n",
      "done with NZ_FQXZ01000030.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002529621124267578\n",
      "Till 50 run took 0.013276338577270508\n",
      "Till 100 run took 0.013892173767089844\n",
      "Till 150 run took 0.013802051544189453\n",
      "Till 200 run took 0.015419721603393555\n",
      "Till 250 run took 0.01562976837158203\n",
      "Till 300 run took 0.015971899032592773\n",
      "Till 350 run took 0.015984058380126953\n",
      "done with NZ_FQXZ01000015.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0005283355712890625\n",
      "done with NZ_FQXZ01000034.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003294944763183594\n",
      "Till 50 run took 0.015207052230834961\n",
      "done with NZ_FQXZ01000012.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00032830238342285156\n",
      "Till 50 run took 0.015586614608764648\n",
      "Till 100 run took 0.013768672943115234\n",
      "done with NZ_FQXZ01000044.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00028634071350097656\n",
      "done with NZ_FQXZ01000055.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00027108192443847656\n",
      "done with NZ_FQXZ01000008.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00028705596923828125\n",
      "Till 50 run took 0.01564335823059082\n",
      "Till 100 run took 0.015352487564086914\n",
      "done with NZ_FQXZ01000021.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002982616424560547\n",
      "Till 50 run took 0.015698671340942383\n",
      "Till 100 run took 0.015546560287475586\n",
      "done with NZ_FQXZ01000009.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00031065940856933594\n",
      "done with NZ_FQXZ01000024.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002875328063964844\n",
      "done with NZ_FQXZ01000025.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003247261047363281\n",
      "Till 50 run took 0.015047788619995117\n",
      "Till 100 run took 0.015151262283325195\n",
      "done with NZ_FQXZ01000029.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003101825714111328\n",
      "Till 50 run took 0.014868497848510742\n",
      "Till 100 run took 0.014949560165405273\n",
      "done with NZ_FQXZ01000037.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00026345252990722656\n",
      "done with NZ_FQXZ01000043.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00024080276489257812\n",
      "done with NZ_FQXZ01000057.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00024580955505371094\n",
      "Till 50 run took 0.012677192687988281\n",
      "done with NZ_FQXZ01000019.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00032329559326171875\n",
      "Till 50 run took 0.015047788619995117\n",
      "done with NZ_FQXZ01000033.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003082752227783203\n",
      "done with NZ_FQXZ01000065.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002646446228027344\n",
      "Till 50 run took 0.014525175094604492\n",
      "Till 100 run took 0.015131473541259766\n",
      "done with NZ_FQXZ01000036.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003275871276855469\n",
      "done with NZ_FQXZ01000051.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003077983856201172\n",
      "done with NZ_FQXZ01000038.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0003247261047363281\n",
      "Till 50 run took 0.015665531158447266\n",
      "done with NZ_FQXZ01000040.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.00029969215393066406\n",
      "done with NZ_FQXZ01000042.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "done with NZ_FQXZ01000075.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n",
      "Till 0 run took 0.0002944469451904297\n",
      "Till 50 run took 0.014784097671508789\n",
      "Till 100 run took 0.01518559455871582\n",
      "done with NZ_FQXZ01000045.1 Vibrio aerogenes CECT 7868, whole genome shotgun sequence\n"
     ]
    }
   ],
   "source": [
    "for header in genomeFragmentDict.keys():\n",
    "    genomeFragmentList = genomeFragmentDict[header]\n",
    "    tetramerDicRepository = {key:[] for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "    begin = time.time()\n",
    "    for n, fragment in enumerate(genomeFragmentList):\n",
    "        tetramerDic = {key:0 for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "        #print('hello')\n",
    "        for pos in range(0,len(fragment)-3):\n",
    "            tetramer = fragment[pos:pos+4]  \n",
    "            #print('hello2')\n",
    "            try:\n",
    "                tetramerDic[tetramer] += 1\n",
    "            except KeyError:\n",
    "                pass\n",
    "\n",
    "        for key, value in tetramerDic.items():\n",
    "            tetramerDicRepository[key].append(value)\n",
    "\n",
    "        if n%50 == 0:\n",
    "            end = time.time()\n",
    "            print(f'Till {n} run took {end-begin}')\n",
    "            begin = time.time()\n",
    "            \n",
    "    print(f'done with {header}')\n",
    "    genomeFragmentDict[header] = pd.DataFrame(data=tetramerDicRepository)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(fileOutput, 'wb') as tetramerFile:\n",
    "    pickle.dump(genomeFragmentDict, tetramerFile)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## V1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">NC_007795.1 Staphylococcus aureus subsp. aureus NCTC 8325 chromosome, complete genome\n",
      "\n",
      ">NZ_CP043554.1 VIBRIO CHOLERAE STRAIN RFB16 CHROMOSOME 1, COMPLETE SEQUENCE\n",
      "2821\n"
     ]
    }
   ],
   "source": [
    "lenFragment = 1000\n",
    "genomeFragmentList = []\n",
    "fileName = 'vibrio_staph_together.fna'#\"GCF_000013425.1_ASM1342v1_genomic.fna(staph aureus)\"#\"Hodgkinia_cicadicola_tetund1_genome.fasta\"#\"Staph_argenteus_58113_genome.fasta\"\"Deinococcus_radiodurans_R1dM1_genome.fasta\"#\n",
    "with open(fileName) as file:\n",
    "    fragment = \"\"\n",
    "    print(file.readline())\n",
    "    for line in file.readlines():\n",
    "        line = line.rstrip().upper()\n",
    "        #print(line)\n",
    "        if len(fragment) < lenFragment:\n",
    "            pass\n",
    "        else:\n",
    "            genomeFragmentList.append(fragment[:lenFragment])\n",
    "            fragment = fragment[lenFragment:]\n",
    "        \n",
    "        try:\n",
    "            if line[0] == '>':\n",
    "                print(line)\n",
    "                break\n",
    "            else:\n",
    "                fragment += line\n",
    "        except IndexError:\n",
    "            print(line)\n",
    "print(len(genomeFragmentList))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "exampleSequence = 'AGGGCTTGCTATAT'\n",
    "tetramerDicRepository = {key:[] for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "tetramerDic = {key:0 for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "import time"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hello1\n",
      "Till 0 run took 0.0010313987731933594\n",
      "Till 50 run took 0.014702558517456055\n",
      "Till 100 run took 0.0175631046295166\n",
      "Till 150 run took 0.016111373901367188\n",
      "Till 200 run took 0.01566910743713379\n",
      "Till 250 run took 0.0161435604095459\n",
      "Till 300 run took 0.01519918441772461\n",
      "Till 350 run took 0.015215396881103516\n",
      "Till 400 run took 0.015461921691894531\n",
      "Till 450 run took 0.014884471893310547\n",
      "Till 500 run took 0.015148639678955078\n",
      "Till 550 run took 0.01526188850402832\n",
      "Till 600 run took 0.014828205108642578\n",
      "Till 650 run took 0.014682531356811523\n",
      "Till 700 run took 0.014548063278198242\n",
      "Till 750 run took 0.01456141471862793\n",
      "Till 800 run took 0.015363931655883789\n",
      "Till 850 run took 0.015027761459350586\n",
      "Till 900 run took 0.013454675674438477\n",
      "Till 950 run took 0.013785600662231445\n",
      "Till 1000 run took 0.015024185180664062\n",
      "Till 1050 run took 0.015482187271118164\n",
      "Till 1100 run took 0.015286922454833984\n",
      "Till 1150 run took 0.014827728271484375\n",
      "Till 1200 run took 0.015070915222167969\n",
      "Till 1250 run took 0.015491962432861328\n",
      "Till 1300 run took 0.01508021354675293\n",
      "Till 1350 run took 0.01491999626159668\n",
      "Till 1400 run took 0.014718294143676758\n",
      "Till 1450 run took 0.014295578002929688\n",
      "Till 1500 run took 0.012501955032348633\n",
      "Till 1550 run took 0.013233661651611328\n",
      "Till 1600 run took 0.013858556747436523\n",
      "Till 1650 run took 0.014838933944702148\n",
      "Till 1700 run took 0.01555776596069336\n",
      "Till 1750 run took 0.015414953231811523\n",
      "Till 1800 run took 0.015338897705078125\n",
      "Till 1850 run took 0.014833688735961914\n",
      "Till 1900 run took 0.01332998275756836\n",
      "Till 1950 run took 0.012862682342529297\n",
      "Till 2000 run took 0.012348651885986328\n",
      "Till 2050 run took 0.012779712677001953\n",
      "Till 2100 run took 0.012389183044433594\n",
      "Till 2150 run took 0.012460470199584961\n",
      "Till 2200 run took 0.012682914733886719\n",
      "Till 2250 run took 0.012346267700195312\n",
      "Till 2300 run took 0.013004064559936523\n",
      "Till 2350 run took 0.013231039047241211\n",
      "Till 2400 run took 0.013104915618896484\n",
      "Till 2450 run took 0.01312398910522461\n",
      "Till 2500 run took 0.013827800750732422\n",
      "Till 2550 run took 0.014938116073608398\n",
      "Till 2600 run took 0.015679359436035156\n",
      "Till 2650 run took 0.015062093734741211\n",
      "Till 2700 run took 0.014932870864868164\n",
      "Till 2750 run took 0.014651298522949219\n",
      "Till 2800 run took 0.01484990119934082\n",
      "Till 2850 run took 0.014902591705322266\n",
      "Till 2900 run took 0.013920068740844727\n",
      "done\n"
     ]
    }
   ],
   "source": [
    "begin = time.time()\n",
    "print('hello1')\n",
    "for n, fragment in enumerate(genomeFragmentList):\n",
    "    tetramerDic = {key:0 for key in ['AAAA', 'AAAC', 'AAAG', 'AAAT', 'AACA', 'AACC', 'AACG', 'AACT', 'AAGA', 'AAGC', 'AAGG', 'AAGT', 'AATA', 'AATC', 'AATG', 'AATT', 'ACAA', 'ACAC', 'ACAG', 'ACAT', 'ACCA', 'ACCC', 'ACCG', 'ACCT', 'ACGA', 'ACGC', 'ACGG', 'ACGT', 'ACTA', 'ACTC', 'ACTG', 'ACTT', 'AGAA', 'AGAC', 'AGAG', 'AGAT', 'AGCA', 'AGCC', 'AGCG', 'AGCT', 'AGGA', 'AGGC', 'AGGG', 'AGGT', 'AGTA', 'AGTC', 'AGTG', 'AGTT', 'ATAA', 'ATAC', 'ATAG', 'ATAT', 'ATCA', 'ATCC', 'ATCG', 'ATCT', 'ATGA', 'ATGC', 'ATGG', 'ATGT', 'ATTA', 'ATTC', 'ATTG', 'ATTT', 'CAAA', 'CAAC', 'CAAG', 'CAAT', 'CACA', 'CACC', 'CACG', 'CACT', 'CAGA', 'CAGC', 'CAGG', 'CAGT', 'CATA', 'CATC', 'CATG', 'CATT', 'CCAA', 'CCAC', 'CCAG', 'CCAT', 'CCCA', 'CCCC', 'CCCG', 'CCCT', 'CCGA', 'CCGC', 'CCGG', 'CCGT', 'CCTA', 'CCTC', 'CCTG', 'CCTT', 'CGAA', 'CGAC', 'CGAG', 'CGAT', 'CGCA', 'CGCC', 'CGCG', 'CGCT', 'CGGA', 'CGGC', 'CGGG', 'CGGT', 'CGTA', 'CGTC', 'CGTG', 'CGTT', 'CTAA', 'CTAC', 'CTAG', 'CTAT', 'CTCA', 'CTCC', 'CTCG', 'CTCT', 'CTGA', 'CTGC', 'CTGG', 'CTGT', 'CTTA', 'CTTC', 'CTTG', 'CTTT', 'GAAA', 'GAAC', 'GAAG', 'GAAT', 'GACA', 'GACC', 'GACG', 'GACT', 'GAGA', 'GAGC', 'GAGG', 'GAGT', 'GATA', 'GATC', 'GATG', 'GATT', 'GCAA', 'GCAC', 'GCAG', 'GCAT', 'GCCA', 'GCCC', 'GCCG', 'GCCT', 'GCGA', 'GCGC', 'GCGG', 'GCGT', 'GCTA', 'GCTC', 'GCTG', 'GCTT', 'GGAA', 'GGAC', 'GGAG', 'GGAT', 'GGCA', 'GGCC', 'GGCG', 'GGCT', 'GGGA', 'GGGC', 'GGGG', 'GGGT', 'GGTA', 'GGTC', 'GGTG', 'GGTT', 'GTAA', 'GTAC', 'GTAG', 'GTAT', 'GTCA', 'GTCC', 'GTCG', 'GTCT', 'GTGA', 'GTGC', 'GTGG', 'GTGT', 'GTTA', 'GTTC', 'GTTG', 'GTTT', 'TAAA', 'TAAC', 'TAAG', 'TAAT', 'TACA', 'TACC', 'TACG', 'TACT', 'TAGA', 'TAGC', 'TAGG', 'TAGT', 'TATA', 'TATC', 'TATG', 'TATT', 'TCAA', 'TCAC', 'TCAG', 'TCAT', 'TCCA', 'TCCC', 'TCCG', 'TCCT', 'TCGA', 'TCGC', 'TCGG', 'TCGT', 'TCTA', 'TCTC', 'TCTG', 'TCTT', 'TGAA', 'TGAC', 'TGAG', 'TGAT', 'TGCA', 'TGCC', 'TGCG', 'TGCT', 'TGGA', 'TGGC', 'TGGG', 'TGGT', 'TGTA', 'TGTC', 'TGTG', 'TGTT', 'TTAA', 'TTAC', 'TTAG', 'TTAT', 'TTCA', 'TTCC', 'TTCG', 'TTCT', 'TTGA', 'TTGC', 'TTGG', 'TTGT', 'TTTA', 'TTTC', 'TTTG', 'TTTT']}\n",
    "    #print('hello')\n",
    "    for pos in range(0,len(fragment)-3):\n",
    "        tetramer = fragment[pos:pos+4]  \n",
    "        #print('hello2')\n",
    "        try:\n",
    "            tetramerDic[tetramer] += 1\n",
    "        except KeyError:\n",
    "            pass\n",
    "\n",
    "    for key, value in tetramerDic.items():\n",
    "        tetramerDicRepository[key].append(value)\n",
    "\n",
    "    if n%50 == 0:\n",
    "        end = time.time()\n",
    "        print(f'Till {n} run took {end-begin}')\n",
    "        begin = time.time()\n",
    "print('done')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[4, 4, 11, 7, 1, 6, 4, 4, 2, 4, 2, 7, 8, 4, 5, 4, 1, 7, 8, 5, 5, 7, 12, 4, 6, 14, 11, 9, 9, 8, 11, 5, 6, 8, 4, 5, 5, 6, 11, 7, 5, 3, 3, 8, 5, 7, 5, 4, 9, 10, 3, 7, 4, 3, 7, 6, 4, 4, 3, 1, 3, 6, 2, 7, 5, 8, 4, 5, 7, 5, 2, 6, 5, 8, 4, 8, 5, 3, 8, 5, 1, 9, 20, 7, 6, 12, 3, 4, 5, 8, 4, 5, 6, 5, 8, 4, 6, 2, 10, 7, 1, 7, 5, 7, 9, 5, 2, 7, 3, 5, 10, 6, 3, 4, 5, 5, 9, 8, 6, 2, 2, 3, 6, 5, 2, 6, 5, 5, 6, 4, 7, 5, 5, 6, 5, 5, 4, 5, 3, 6, 7, 3, 2, 3, 0, 3, 4, 4, 4, 5, 6, 7, 8, 4, 4, 2, 5, 3, 4, 7, 3, 2, 14, 8, 6, 8, 5, 2, 3, 5, 5, 1, 7, 12, 4, 8, 9, 2, 5, 7, 7, 9, 1, 8, 9, 3, 5, 4, 5, 8, 4, 7, 7, 4, 8, 8, 4, 6, 5, 5, 8, 4, 2, 8, 3, 11, 7, 5, 5, 6, 0, 8, 7, 4, 11, 1, 4, 6, 6, 7, 3, 5, 5, 6, 9, 10, 6, 6, 10, 2, 7, 4, 4, 6, 6, 7, 4, 6, 7, 3, 5, 6, 7, 4, 5, 7, 6, 5, 4, 3, 5, 4, 4, 4, 8, 6, 8, 4, 2, 3, 2, 9, 3, 2, 11, 3, 8, 10, 6, 8, 11, 11, 7, 4, 3, 5, 3, 11, 6, 3, 6, 5, 5, 8, 6, 5, 6, 7, 7, 3, 10, 5, 7, 5, 6, 3, 7, 4, 5, 2, 4, 7, 6, 2, 6, 2, 3, 5, 1, 4, 3, 4, 3, 2, 3, 4, 5, 1, 5, 7, 1, 6, 3, 3, 3, 4, 4, 3, 2, 4, 4, 8, 5, 7, 3, 6, 6, 5, 3, 2, 6, 7, 2, 4, 4, 5, 5, 2, 5, 3, 2, 4, 8, 5, 5, 5, 5, 9, 5, 3, 5, 3, 4, 4, 6, 5, 5, 5, 6, 5, 6, 7, 8, 5, 11, 4, 8, 11, 4, 3, 5, 6, 5, 6, 4, 3, 3, 10, 4, 11, 9, 7, 8, 8, 8, 3, 4, 3, 7, 3, 4, 10, 5, 9, 8, 9, 4, 5, 5, 8, 10, 9, 6, 6, 9, 3, 7, 9, 4, 7, 4, 7, 3, 4, 3, 6, 4, 3, 3, 6, 2, 6, 5, 7, 7, 9, 9, 9, 5, 6, 3, 2, 8, 8, 5, 9, 2, 5, 6, 9, 4, 3, 4, 7, 8, 5, 8, 5, 4, 7, 9, 1, 3, 8, 4, 4, 6, 5, 4, 7, 5, 6, 9, 7, 7, 7, 6, 2, 10, 7, 6, 6, 5, 4, 9, 3, 3, 8, 7, 9, 8, 8, 3, 9, 5, 5, 4, 4, 7, 6, 5, 4, 4, 4, 4, 7, 8, 5, 4, 3, 6, 7, 2, 5, 6, 3, 7, 3, 8, 8, 3, 5, 6, 4, 7, 9, 9, 5, 3, 5, 8, 10, 8, 10, 2, 8, 4, 9, 6, 6, 9, 6, 15, 3, 6, 4, 3, 5, 9, 9, 7, 6, 3, 2, 3, 2, 11, 0, 3, 10, 1, 5, 4, 4, 2, 2, 6, 6, 6, 5, 8, 6, 6, 4, 7, 8, 5, 2, 8, 7, 5, 5, 7, 5, 3, 1, 8, 4, 2, 7, 6, 6, 7, 4, 2, 1, 4, 7, 2, 2, 5, 1, 4, 6, 7, 6, 8, 5, 7, 5, 6, 5, 1, 4, 2, 7, 3, 4, 6, 8, 4, 7, 4, 4, 4, 7, 5, 6, 5, 5, 4, 9, 4, 5, 4, 6, 5, 7, 4, 2, 6, 1, 3, 6, 5, 7, 3, 4, 6, 7, 5, 7, 5, 11, 1, 8, 5, 5, 2, 4, 4, 7, 6, 4, 5, 4, 3, 5, 10, 5, 9, 6, 8, 5, 6, 5, 6, 8, 6, 4, 2, 6, 3, 3, 6, 3, 6, 5, 4, 8, 8, 7, 6, 4, 9, 4, 5, 5, 4, 10, 10, 8, 3, 8, 5, 6, 8, 4, 8, 5, 5, 5, 5, 9, 4, 6, 5, 4, 3, 7, 4, 1, 3, 4, 7, 7, 4, 3, 8, 4, 4, 9, 5, 6, 8, 10, 8, 12, 4, 2, 6, 5, 3, 6, 5, 7, 9, 5, 5, 10, 4, 13, 5, 7, 8, 5, 9, 2, 2, 5, 6, 8, 8, 6, 6, 3, 9, 7, 4, 6, 10, 8, 9, 7, 2, 5, 5, 4, 9, 3, 1, 5, 0, 9, 8, 5, 7, 4, 9, 6, 6, 11, 9, 4, 7, 6, 4, 10, 9, 5, 6, 8, 5, 6, 8, 8, 4, 10, 1, 3, 10, 7, 5, 8, 8, 6, 7, 2, 6, 7, 8, 5, 4, 6, 3, 8, 9, 8, 1, 6, 6, 7, 5, 6, 3, 8, 9, 9, 9, 4, 8, 6, 5, 5, 8, 13, 2, 8, 7, 9, 0, 3, 3, 3, 9, 5, 6, 10, 8, 8, 3, 2, 4, 7, 4, 9, 7, 5, 7, 8, 5, 4, 2, 3, 5, 6, 1, 10, 4, 3, 2, 6, 7, 4, 2, 6, 3, 8, 3, 11, 7, 7, 2, 1, 6, 5, 7, 2, 9, 9, 3, 4, 4, 4, 6, 1, 5, 3, 7, 6, 6, 7, 1, 4, 2, 6, 9, 6, 7, 8, 3, 4, 7, 4, 6, 4, 3, 5, 5, 4, 9, 4, 6, 7, 8, 4, 7, 9, 3, 6, 2, 4, 7, 5, 10, 11, 2, 4, 6, 6, 5, 6, 6, 5, 5, 9, 4, 7, 11, 10, 10, 4, 8, 7, 4, 6, 8, 3, 5, 9, 10, 7, 4, 3, 3, 8, 7, 6, 7, 5, 6, 5, 7, 7, 5, 7, 5, 3, 2, 9, 5, 10, 7, 13, 11, 6, 1, 3, 1, 9, 9, 4, 7, 7, 9, 6, 10, 4, 7, 4, 4, 6, 4, 7, 6, 3, 6, 1, 11, 3, 2, 5, 11, 1, 2, 4, 5, 5, 5, 2, 12, 4, 8, 6, 6, 5, 1, 6, 7, 5, 6, 6, 2, 6, 8, 6, 4, 6, 5, 7, 8, 5, 9, 6, 4, 5, 9, 8, 4, 5, 7, 7, 2, 4, 7, 4, 4, 10, 5, 5, 4, 1, 4, 7, 8, 1, 6, 6, 4, 7, 7, 3, 4, 6, 2, 3, 6, 5, 6, 7, 4, 6, 6, 9, 2, 2, 4, 5, 5, 5, 8, 6, 2, 5, 6, 4, 4, 5, 6, 2, 7, 5, 5, 6, 5, 6, 2, 2, 3, 5, 4, 7, 9, 6, 8, 7, 10, 10, 6, 6, 15, 19, 11, 6, 6, 5, 8, 11, 9, 5, 12, 2, 4, 10, 7, 7, 13, 2, 0, 8, 7, 3, 9, 2, 4, 7, 11, 6, 5, 5, 8, 4, 5, 6, 9, 7, 6, 8, 2, 7, 6, 2, 7, 6, 3, 7, 7, 8, 6, 6, 4, 3, 4, 5, 7, 9, 9, 8, 7, 7, 5, 2, 11, 10, 6, 2, 6, 4, 5, 4, 3, 2, 4, 4, 10, 8, 3, 6, 12, 7, 7, 6, 8, 11, 6, 10, 7, 8, 4, 3, 6, 4, 15, 3, 2, 5, 3, 4, 6, 3, 5, 2, 5, 5, 10, 6, 9, 9, 3, 5, 5, 3, 4, 5, 7, 8, 13, 3, 6, 7, 7, 13, 6, 8, 8, 6, 14, 3, 11, 3, 7, 5, 3, 6, 2, 6, 7, 6, 8, 3, 7, 7, 4, 5, 4, 11, 6, 5, 7, 1, 4, 8, 7, 4, 6, 10, 6, 9, 3, 7, 8, 0, 3, 0, 2, 6, 2, 6, 5, 9, 2, 8, 9, 11, 6, 7, 5, 6, 6, 2, 2, 3, 2, 3, 3, 7, 10, 2, 4, 3, 7, 9, 5, 11, 4, 4, 8, 9, 14, 6, 5, 0, 7, 3, 9, 8, 4, 7, 4, 16, 10, 4, 5, 8, 6, 6, 7, 8, 7, 9, 5, 6, 6, 4, 4, 1, 9, 7, 7, 5, 5, 8, 3, 1, 6, 2, 4, 6, 7, 8, 4, 1, 2, 5, 3, 5, 2, 4, 3, 12, 4, 7, 3, 2, 10, 14, 4, 9, 9, 6, 5, 10, 9, 8, 3, 13, 8, 7, 8, 8, 5, 14, 11, 6, 6, 7, 8, 6, 8, 5, 5, 7, 10, 5, 13, 6, 9, 8, 4, 10, 4, 14, 11, 7, 5, 7, 7, 6, 5, 8, 5, 12, 3, 10, 7, 11, 7, 10, 9, 9, 5, 4, 6, 9, 13, 8, 4, 4, 8, 6, 9, 6, 5, 8, 8, 5, 10, 10, 8, 8, 9, 4, 10, 5, 5, 5, 4, 5, 5, 4, 5, 0, 3, 5, 5, 3, 6, 10, 5, 8, 5, 8, 7, 2, 9, 7, 4, 8, 4, 6, 5, 7, 5, 5, 5, 6, 9, 8, 6, 7, 7, 6, 13, 8, 3, 3, 7, 8, 5, 12, 9, 10, 5, 4, 7, 6, 12, 7, 8, 10, 6, 8, 9, 6, 5, 7, 4, 4, 12, 12, 5, 3, 6, 9, 4, 7, 8, 10, 8, 7, 5, 8, 6, 4, 7, 5, 7, 4, 7, 7, 4, 10, 6, 5, 12, 10, 3, 7, 6, 3, 12, 6, 12, 10, 7, 5, 12, 5, 11, 6, 4, 6, 8, 6, 6, 4, 3, 9, 10, 3, 9, 7, 6, 11, 1, 6, 9, 11, 7, 10, 12, 5, 5, 8, 11, 11, 10, 7, 4, 6, 7, 5, 4, 5, 4, 6, 5, 6, 5, 4, 6, 10, 11, 7, 5, 4, 3, 2, 8, 5, 6, 8, 9, 8, 8, 7, 8, 8, 7, 8, 10, 6, 5, 5, 4, 8, 4, 2, 10, 8, 7, 5, 4, 4, 8, 3, 6, 9, 10, 9, 6, 9, 7, 4, 4, 6, 10, 7, 6, 5, 2, 12, 5, 9, 2, 8, 5, 4, 5, 8, 6, 9, 8, 12, 11, 8, 4, 8, 15, 6, 7, 2, 9, 3, 10, 7, 5, 4, 7, 6, 6, 7, 10, 8, 8, 4, 3, 4, 4, 7, 7, 6, 5, 10, 9, 4, 2, 7, 6, 5, 11, 10, 6, 4, 8, 14, 9, 6, 7, 5, 6, 3, 5, 8, 2, 8, 9, 7, 6, 5, 6, 1, 10, 5, 8, 8, 4, 7, 9, 11, 11, 10, 13, 5, 7, 8, 3, 4, 5, 5, 6, 3, 7, 7, 13, 11, 6, 6, 9, 8, 11, 15, 8, 6, 10, 4, 5, 6, 4, 7, 8, 6, 9, 8, 7, 11, 6, 5, 9, 5, 3, 11, 1, 10, 7, 12, 3, 7, 4, 5, 4, 5, 4, 12, 7, 6, 3, 6, 9, 6, 7, 6, 9, 6, 4, 5, 6, 7, 7, 4, 6, 4, 3, 7, 6, 7, 7, 3, 10, 4, 6, 5, 7, 7, 3, 4, 4, 5, 8, 11, 4, 6, 4, 3, 6, 10, 8, 5, 11, 7, 10, 11, 5, 7, 7, 4, 5, 11, 4, 5, 5, 6, 11, 9, 14, 13, 5, 6, 5, 9, 2, 4, 8, 6, 8, 5, 15, 5, 4, 13, 5, 6, 4, 5, 8, 6, 8, 3, 3, 7, 6, 7, 10, 10, 3, 5, 5, 4, 9, 11, 6, 7, 10, 9, 6, 5, 12, 8, 7, 12, 5, 8, 8, 5, 3, 1, 10, 10, 8, 10, 8, 12, 8, 5, 13, 6, 7, 3, 9, 7, 7, 5, 10, 6, 4, 6, 9, 12, 6, 5, 6, 3, 3, 4, 9, 10, 5, 3, 7, 5, 8, 8, 6, 6, 7, 9, 6, 5, 11, 4, 9, 11, 7, 3, 6, 11, 4, 6, 10, 2, 2, 1, 3, 2, 7, 5, 9, 9, 9, 8, 7, 6, 8, 7, 6, 8, 5, 5, 5, 6, 8, 9, 4, 10, 5, 9, 11, 5, 4, 5, 7, 12, 6, 8, 7, 5, 6, 9, 4, 11, 1, 9, 3, 5, 3, 1, 6, 10, 6, 9, 9, 7, 8, 7, 2, 6, 10, 8, 6, 6, 7, 4, 9, 7, 4, 7, 4, 7, 8, 13, 9, 5, 6, 6, 7, 7, 10, 8, 8, 5, 11, 3, 7, 11, 13, 13, 4, 5, 9, 7, 2, 10, 11, 6, 3, 11, 7, 7, 11, 4, 4, 4, 6, 12, 9, 7, 8, 5, 13, 2, 4, 8, 12, 8, 8, 6, 3, 9, 8, 6, 3, 8, 8, 7, 5, 8, 4, 6, 9, 9, 9, 7, 7, 6, 4, 5, 8, 4, 10, 5, 6, 16, 3, 9, 6, 3, 5, 2, 5, 8, 13, 6, 9, 12, 5, 6, 13, 5, 10, 7, 11, 13, 5, 8, 1, 6, 11, 5, 7, 5, 5, 7, 3, 7, 7, 10, 10, 6, 6, 4, 13, 8, 10, 9, 6, 6, 7, 7, 10, 12, 5, 6, 6, 10, 2, 3, 1, 13, 3, 4, 7, 8, 6, 6, 8, 6, 5, 4, 6, 7, 6, 6, 2, 4, 9, 4, 3, 8, 10, 4, 5, 8, 8, 3, 5, 5, 5, 7, 10, 5, 3, 5, 7, 6, 6, 5, 7, 3, 7, 7, 6, 13, 8, 6, 11, 6, 12, 11, 10, 4, 3, 8, 12, 6, 7, 11, 4, 7, 9, 6, 3, 1, 8, 4, 4, 4, 0, 8, 6, 7, 9, 11, 9, 8, 8, 6, 6, 9, 4, 3, 7, 10, 5, 9, 7, 5, 7, 4, 6, 11, 11, 7, 7, 6, 6, 6, 9, 9, 5, 16, 13, 10, 8, 12, 6, 12, 12, 6, 6, 10, 10, 8, 4, 8, 3, 4, 2, 7, 10, 6, 5, 5, 6, 3, 8, 9, 6, 8, 7, 10, 3, 8, 13, 10, 3, 5, 8, 6, 4, 1, 5, 2, 4, 11, 5, 5, 7, 6, 5, 7, 4, 8, 8, 10, 5, 5, 7, 10, 10, 3, 4, 4, 3, 3, 2, 9, 10, 5, 11, 8, 7, 8, 4, 5, 9, 5, 6, 4, 8, 4, 4, 5, 6, 7, 5, 8, 11, 5, 13, 5, 6, 6, 6, 10, 5, 4, 5, 8, 8, 3, 6, 1, 9, 3, 5, 4, 8, 6, 9, 10, 8, 7, 11, 5, 9, 3, 11, 11, 1, 7, 7, 7, 6, 5, 3, 6, 10, 5, 3, 2, 10, 6, 9, 5, 4, 6, 8, 3, 7, 5, 12, 5, 10, 4, 5, 3, 9, 4, 5, 4, 10, 8, 5, 8, 11, 12, 13, 6, 4, 9, 8, 4, 3, 3, 1, 8, 2, 7, 10, 6, 5, 7, 11, 9, 3, 5, 5, 3, 4, 1, 3, 9, 1, 7, 7, 8, 10, 3, 5, 1, 7, 5, 2, 5, 5, 2, 8, 7, 3, 6, 13, 8, 7, 5, 6, 6, 8, 7, 4, 4, 5, 3, 3, 4, 6, 7, 6, 10, 5, 14, 2, 12, 12, 7, 4, 10, 4, 1, 8, 5, 6, 6, 8, 10, 5, 8, 5, 12, 9, 10, 6, 6, 10, 10, 8, 8, 9, 12, 6, 11, 8, 9, 7, 16, 3, 5, 5, 1, 0, 5, 9, 7, 8, 5, 3, 8, 5, 6, 6, 1, 4, 5, 4, 5, 6, 8, 10, 6, 8, 4, 8, 7, 8, 11, 9, 10, 5, 3, 4, 4, 3, 2, 4, 8, 8, 9, 7, 9, 5, 7, 4, 6, 12, 5, 7, 6, 4, 9, 2, 5, 8, 4, 2, 2, 9, 2, 5, 4, 6, 2, 4, 11, 5, 5, 7, 7, 9, 6, 7, 2, 2, 6, 4, 5, 7, 3, 6, 1, 5, 10, 4, 1, 3, 3, 5, 6, 6, 4, 7, 2, 4, 4, 8, 6, 7, 3, 1, 6, 5, 4, 4, 5, 6, 5, 8, 5, 7, 7, 6, 2, 6, 5, 9, 8, 4, 6, 4, 5, 10, 4, 5, 3, 7, 7, 6, 17, 4, 5, 6, 6, 4, 7, 0, 4, 3, 5, 0, 6, 3, 6, 3, 3, 6, 0, 2, 2, 6, 10, 3, 8, 6, 3, 1, 8, 4, 2, 6, 2, 1, 9, 2, 3, 1, 8, 9, 7, 6, 7, 5, 5, 9, 10, 14, 17, 15, 7, 3, 13, 9, 6, 8, 7, 8, 6, 6, 8, 6, 6, 4, 9, 12, 8, 4, 6, 6, 5, 7, 4, 10, 4, 1, 2, 6, 20, 8, 7, 10, 10, 3, 5, 12, 5, 8, 9, 6, 6, 10, 8, 10, 13, 13, 10, 6, 13, 6, 6, 6, 11, 12, 7, 5, 5, 8, 5, 11, 4, 7, 1, 3, 4, 6, 4, 10, 4, 8, 7, 3, 14, 7, 5, 6, 5, 4, 6, 5, 11, 6, 5, 4, 7, 6, 5, 8, 11, 12, 6, 9, 4, 9, 8, 8, 9, 5, 7, 7, 7, 9, 8, 7, 4, 10, 8, 4, 7, 2, 7, 9, 2, 7, 3, 4, 7, 6, 9, 2, 7, 9, 12, 4, 11, 10, 11, 10, 9, 8, 5, 10, 5, 11, 7, 6, 5, 8, 2, 9, 5, 8, 2, 4, 8, 8, 4, 7, 5, 7, 8, 10, 4, 9, 4, 10, 6, 6, 7, 4, 6, 3, 1, 6, 7, 11, 4, 6, 2, 0, 9, 3, 4, 5, 1, 4, 4, 2, 5, 9, 10, 13, 5, 4, 7, 9, 7, 4, 6, 7, 9, 6, 1, 4, 9, 9, 5, 7, 4, 3, 1, 4, 2, 7, 7, 7, 12, 7, 5, 3, 4, 6, 6, 5, 5, 9, 6, 7, 6, 1, 4, 4, 3, 7, 5, 3, 3, 3, 4, 5, 6, 4, 5, 7, 7, 8, 2, 3, 4, 9, 6, 5, 6, 7, 8, 7, 6, 7, 5, 7, 5, 9, 6]\n"
     ]
    }
   ],
   "source": [
    "print(tetramerDicRepository['AAAC'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# on command line write\n",
    "#     pip install pandas\n",
    "\n",
    "# Write this and give it your dictionary (hope the shape is as expected)\n",
    "import pandas as pd\n",
    "tetramerDic = tetramerDicRepository\n",
    "# should be shaped like\n",
    "#     {'AAAA':[num,num,num...], \n",
    "#      'AAAT':[num,num,num...], ... }\n",
    "\n",
    "df = pd.DataFrame(data=tetramerDic)\n",
    "#df.to_csv('give_to_matthew_plz.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAA</th>\n",
       "      <th>AAAC</th>\n",
       "      <th>AAAG</th>\n",
       "      <th>AAAT</th>\n",
       "      <th>AACA</th>\n",
       "      <th>AACC</th>\n",
       "      <th>AACG</th>\n",
       "      <th>AACT</th>\n",
       "      <th>AAGA</th>\n",
       "      <th>AAGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TTCG</th>\n",
       "      <th>TTCT</th>\n",
       "      <th>TTGA</th>\n",
       "      <th>TTGC</th>\n",
       "      <th>TTGG</th>\n",
       "      <th>TTGT</th>\n",
       "      <th>TTTA</th>\n",
       "      <th>TTTC</th>\n",
       "      <th>TTTG</th>\n",
       "      <th>TTTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>8</td>\n",
       "      <td>11</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>6</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>5</td>\n",
       "      <td>...</td>\n",
       "      <td>7</td>\n",
       "      <td>0</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>5</td>\n",
       "      <td>6</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>6</td>\n",
       "      <td>11</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>11</td>\n",
       "      <td>13</td>\n",
       "      <td>1</td>\n",
       "      <td>9</td>\n",
       "      <td>5</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>3</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4</td>\n",
       "      <td>6</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>8</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>6</td>\n",
       "      <td>4</td>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>7</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>5</td>\n",
       "      <td>2</td>\n",
       "      <td>10</td>\n",
       "      <td>2</td>\n",
       "      <td>5</td>\n",
       "      <td>3</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   AAAA  AAAC  AAAG  AAAT  AACA  AACC  AACG  AACT  AAGA  AAGC  ...  TTCG  \\\n",
       "0     5     4     7     3     3     4     3     6     8    11  ...     2   \n",
       "1     5     4     4     5     3     6     4     3     5     5  ...     7   \n",
       "2     6    11     7     9    11    13     1     9     5     8  ...     3   \n",
       "3     7     7     7     3     2     4     6     5     3     8  ...     1   \n",
       "4     1     1     6     4     2     3     5     3     7     9  ...     5   \n",
       "\n",
       "   TTCT  TTGA  TTGC  TTGG  TTGT  TTTA  TTTC  TTTG  TTTT  \n",
       "0     1     9     3     3     3     4     2     6     1  \n",
       "1     0     4     5     4     5     6     3     4     2  \n",
       "2     1     5     1     2     8     2     1     4     1  \n",
       "3     3     5     4     2     1     5     0     2     1  \n",
       "4     2    10     2     5     3     4     0     3     0  \n",
       "\n",
       "[5 rows x 256 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "sums = df.sum()\n",
    "df_freq = df.loc[:,'AAAA':'TTTT'].div(sums, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AAAA    24680\n",
       "AAAC    18107\n",
       "AAAG    17536\n",
       "AAAT    17753\n",
       "AACA    14127\n",
       "        ...  \n",
       "TTGT    12271\n",
       "TTTA    15304\n",
       "TTTC    16240\n",
       "TTTG    21112\n",
       "TTTT    24109\n",
       "Length: 256, dtype: int64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sums"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AAAA</th>\n",
       "      <th>AAAC</th>\n",
       "      <th>AAAG</th>\n",
       "      <th>AAAT</th>\n",
       "      <th>AACA</th>\n",
       "      <th>AACC</th>\n",
       "      <th>AACG</th>\n",
       "      <th>AACT</th>\n",
       "      <th>AAGA</th>\n",
       "      <th>AAGC</th>\n",
       "      <th>...</th>\n",
       "      <th>TTCG</th>\n",
       "      <th>TTCT</th>\n",
       "      <th>TTGA</th>\n",
       "      <th>TTGC</th>\n",
       "      <th>TTGG</th>\n",
       "      <th>TTGT</th>\n",
       "      <th>TTTA</th>\n",
       "      <th>TTTC</th>\n",
       "      <th>TTTG</th>\n",
       "      <th>TTTT</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000289</td>\n",
       "      <td>0.000248</td>\n",
       "      <td>0.000498</td>\n",
       "      <td>0.000565</td>\n",
       "      <td>0.000601</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000188</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000459</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>0.000165</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.000203</td>\n",
       "      <td>0.000221</td>\n",
       "      <td>0.000228</td>\n",
       "      <td>0.000282</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000434</td>\n",
       "      <td>0.000330</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.000273</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000657</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000204</td>\n",
       "      <td>0.000275</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.000407</td>\n",
       "      <td>0.000392</td>\n",
       "      <td>0.000185</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.000243</td>\n",
       "      <td>0.000607</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>0.000507</td>\n",
       "      <td>0.000779</td>\n",
       "      <td>0.000940</td>\n",
       "      <td>0.000083</td>\n",
       "      <td>0.000748</td>\n",
       "      <td>0.000353</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000281</td>\n",
       "      <td>0.000082</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000652</td>\n",
       "      <td>0.000131</td>\n",
       "      <td>0.000062</td>\n",
       "      <td>0.000189</td>\n",
       "      <td>0.000041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.000284</td>\n",
       "      <td>0.000387</td>\n",
       "      <td>0.000399</td>\n",
       "      <td>0.000169</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000289</td>\n",
       "      <td>0.000496</td>\n",
       "      <td>0.000415</td>\n",
       "      <td>0.000212</td>\n",
       "      <td>0.000437</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000094</td>\n",
       "      <td>0.000245</td>\n",
       "      <td>0.000255</td>\n",
       "      <td>0.000220</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000081</td>\n",
       "      <td>0.000327</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000095</td>\n",
       "      <td>0.000041</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000055</td>\n",
       "      <td>0.000342</td>\n",
       "      <td>0.000225</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000217</td>\n",
       "      <td>0.000413</td>\n",
       "      <td>0.000249</td>\n",
       "      <td>0.000494</td>\n",
       "      <td>0.000491</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000469</td>\n",
       "      <td>0.000163</td>\n",
       "      <td>0.000510</td>\n",
       "      <td>0.000110</td>\n",
       "      <td>0.000276</td>\n",
       "      <td>0.000244</td>\n",
       "      <td>0.000261</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000142</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 256 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       AAAA      AAAC      AAAG      AAAT      AACA      AACC      AACG  \\\n",
       "0  0.000203  0.000221  0.000399  0.000169  0.000212  0.000289  0.000248   \n",
       "1  0.000203  0.000221  0.000228  0.000282  0.000212  0.000434  0.000330   \n",
       "2  0.000243  0.000607  0.000399  0.000507  0.000779  0.000940  0.000083   \n",
       "3  0.000284  0.000387  0.000399  0.000169  0.000142  0.000289  0.000496   \n",
       "4  0.000041  0.000055  0.000342  0.000225  0.000142  0.000217  0.000413   \n",
       "\n",
       "       AACT      AAGA      AAGC  ...      TTCG      TTCT      TTGA      TTGC  \\\n",
       "0  0.000498  0.000565  0.000601  ...  0.000188  0.000082  0.000459  0.000165   \n",
       "1  0.000249  0.000353  0.000273  ...  0.000657  0.000000  0.000204  0.000275   \n",
       "2  0.000748  0.000353  0.000437  ...  0.000281  0.000082  0.000255  0.000055   \n",
       "3  0.000415  0.000212  0.000437  ...  0.000094  0.000245  0.000255  0.000220   \n",
       "4  0.000249  0.000494  0.000491  ...  0.000469  0.000163  0.000510  0.000110   \n",
       "\n",
       "       TTGG      TTGT      TTTA      TTTC      TTTG      TTTT  \n",
       "0  0.000165  0.000244  0.000261  0.000123  0.000284  0.000041  \n",
       "1  0.000220  0.000407  0.000392  0.000185  0.000189  0.000083  \n",
       "2  0.000110  0.000652  0.000131  0.000062  0.000189  0.000041  \n",
       "3  0.000110  0.000081  0.000327  0.000000  0.000095  0.000041  \n",
       "4  0.000276  0.000244  0.000261  0.000000  0.000142  0.000000  \n",
       "\n",
       "[5 rows x 256 columns]"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_freq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_freq.to_csv('tetraFrequency.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3.9.12 (stats)",
   "language": "python",
   "name": "stats"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
